enhancements/storage/storage-hypershift.md

```diff
@@ -0,0 +1,1050 @@
+  an endless loop. The operators will redeploy the operands with each CR change and the new operands will need to wait
+  for leader election to expire, while the old operands are already terminated.
+  * We mitigate this issue a bit by overwriting user changes in the operator CRs by Hosted Cluster Config Operator that
+    runs in HyperShift management cluster. Still, an evil Cluster Instance Admin can easily write a loop that updates
+   the CRs in a busy loop.
+
+* Since leader election leased of the operands (AWS EBS CSI driver sidecars and csi-snapshot-controller) are in the
+  guest cluster, an evil Cluster Instance Admin can DOS the whole CSI driver / CSI snapshots by holding these leases
+  forever.
+  * To mitigate this we would need to teach upstream CSI sidecars to handle two kubeconfigs. One for leader election
+    in the management cluster and second to handle the actual storage objects in the guest cluster. There is no
+    such plan today upstream.
+
+* Due to our operator design based on library-go controllers, the operators use many conditions and overall
+  ClusterOperator conditions are computed as their union. A Cluster Instance Admin may add a condition
+  to `storage.status`, `clustercsidriver.status` or `csisnapshotcontroller.status` that would be translated by the
+  operators to overall `clusterOperator.status` of the corresponding operator.
+  * For example, adding condition `ExplitDegraded: true` to `storage.status` will render ClusterOperator `storage`
+    `Degraded: true`. cluster-storage-operator will not sanitize it.
+  * Fixing this is hard in library-go, as we use really a lot of conditions and creating a whitelist of the "official
+    ones" would be error-prone.
+
+All three issues are not specific to HyperShift, they exist in standalone OCP too. Fixing all these issues is IMO
+impossible if there is an evil Cluster Instance Admin. If we still want to fix it properly in a future
+OCP release, we may need to redesign storage operators completely! See Alternatives below.
+
+#### Less code sharing
+
+* There is a significant difference between CVO starting cluster-storage-operator and
+  cluster-csi-snapshot-controller-operator in a standalone OCP and control-plane-operator starting the same operator in
+  HyperShift. It would be better if control-plane-operator had ability to process release manifest in a similar fashion
+  as CVO, so we would share the yaml files of our operators in a single place.
+
+#### Code complexity
+
+* Complexity of cluster-storage-operator, cluster-csi-snapshot-controller-operator, and aws-ebs-csi-driver-operator
+  will grow significantly, they need to handle two kubeconfigs / kubeclients where one was needed before. We can expect
+  regressions.
+  * We will mitigate this risk by reworking the unit tests to handle primarily hypershift case. Standalone OCP is
+    a variation of HyperShift, where both kube clients are actually the same and both lead to the standalone OCP.
```

Comment:
This is actually a big risk. Storage is complicated already and this design adds yet another dimension to our support / maintenance matrix. We have 9 CSI driver operators, each of them is a bit different (because the clouds are different!), and keeping track of what operator watches / syncs which object in mgmt/guest cluster is crazy. This is error prone.

Reply:
Is this the first time we've encountered this sort of need for HyperShift? How did we deal with it in other situations?

Reply:
I think storage is the first one to encounter this issue. Other control-plane components are started by code hardcoded to control-plane-operator and IMO they don't try to reuse code from their standalone OCP operators.
(I did not research OLM deployment thoroughly).
@csrwng, is this comment correct?

Reply:
No, the cluster network operator uses two kubeconfigs as well in order to work in a HyperShift environment.